jade/jobs/job_container_by_name.py

```diff
@@ -0,0 +1,49 @@
+"""Implements a container for jobs represented by a key."""
+
+import logging
+
+from jade.exceptions import InvalidParameter
+from jade.jobs.job_container_interface import JobContainerInterface
+
+
+logger = logging.getLogger(__name__)
+
+
+class JobContainerByName(JobContainerInterface):
+    """Stores jobs by name, which must be unique."""
+    def __init__(self):
+        # name: JobParametersInterface
+        self._jobs = {}
+
+    def add_job(self, job):
+        self._jobs[job.name] = job
+        logger.debug("Added job %s", job.name)
+
+    def clear(self):
+        self._jobs.clear()
+        logger.debug("Cleared all jobs.")
+
+    def remove_job(self, job):
+        self._jobs.pop(job.name)
+        logger.info("Removed job %s", job.name)
+
+    def iter_jobs(self):
+        for job in self._jobs.values():
+            yield job
+
+    def get_job(self, name):
+        job = self._jobs.get(name)
+        if job is None:
+            raise InvalidParameter(f"job={name} is not stored")
+        return self._jobs[name]
+
+    def get_jobs(self, sort_by_name=False):
+        if sort_by_name:
+            names = list(self._jobs.keys())
+            names.sort()
+            return [self._jobs[x] for x in names]
+
+        return list(self.iter_jobs())
+
+    def get_num_jobs(self):
+        return len(self._jobs)
```

Comment:
Nothing wrong, no need to change. Here just wants to mention about the magic methods and duck typing in Python, which could make the container-like object more pythonic, like `get_job` vs `__getitem__`, `add_job` vs `__setitem__`, `get_num_jobs` vs `__len__`, etc, more here https://docs.python.org/3/library/collections.abc.html.

Reply:
I thought about this. The original class included methods that accepted a key or a name, so it wasn't exactly like a dict. I didn't want to go change all the existing call sites.

There is still the abstract base class JobContainerInterface which does not require a dict, so I don't think I would want to make the interface behave like a dict. `__iter__` and `__len__` may make sense.

I will admit that I have aversions to doing this kind of thing for a few reasons:
- It makes life very easy for developers. However, it can make things more difficult for some readers.
- You lose some IDE functionality. You can't search for references or jump to definitions.

I suppose we could have both interfaces.